### PR TITLE
perf : Optimize river generation

### PR DESCRIPTION
**Pull request by cogito123**
### Brief description of the changes:
Optimized river generation by reducing amount of API calls, better memory management and lookup table with pseudo random offset values.

#### Current performance:
190x "generate_river" Total Duration: 2949.216225ms
Average time spent in function: 15.5ms

#### With new changes applied (just static buffer)
134x "generate_river" Total Duration: 1359.028429ms
Average time spent in function: 10.1ms

#### With new changes applied (full memory rework)
237x "generate_river" Total Duration: 1754.767363ms
Average time spent in function: 7ms

### Tested Changes:
- [X] I've tested the changes locally.
- [ ]  I've tested the changes online.

### CODE REVIEW

**Comment by Ragnarok77-factorio** [comment]
why someone said it was ugly? confused, I tested it locally, I don't notice anything very different.. 🤔

**Comment by XVhc6A** [comment]
Could you provide a brief description + script that produces these precomputed files?  I'm okay with having them, but if someone wants to change it later it shouldn't be so black-box.

**Comment by Ragnarok77-factorio** [comment]
Tested it, nothing to add, will be tested, and I will review later, I think there is a conflict to be solved though if we wanna test it this saturday. @cogito123

**Comment by developer-8** [comment]
Screenshot from 2023-11-02 00-12-48
1. After this PR, every map has the same coastline pattern
2. The shape of the coastline is different. The original is smoother.

**Comment by developer-8** [comment]
3.
image
Original river variate water/land on 4 tiles. y for north side: -26, -25, -24, -23
image
After this PR it variate on 2 tiles. y for north side: -23, -22
This also means that the river has shrunk by 2 tiles.

4.`generate_river` and as consequence `is_horizontal_border_river` doing work only 1 time for 1 chunk near river so I think its acceptable sacrifice 5.4ms (from 15.5ms to 10.1ms) performance gain and have unique river every game

**Comment by cogito123** [comment]
I'll need to work some more on the pattern to make it smoother. But it's now seeded.

**Comment by cogito123** [comment]
OK, now it's identical to original one, rebased and I also addressed your comment @developer-8. It should be now properly spaced.

**Comment by cogito123** [comment]
> Could you provide a brief description + script that produces these precomputed files? I'm okay with having them, but if someone wants to change it later it shouldn't be so black-box.

I totally agree, current feedback from discord is that I should create separate repo for tool that I use for these precomputed files. At the moment there are several executables for that. I started work on merging all of these into one. Once I have it done, I'll publish it under my name. In the meantime if you propose to do it differently or host it somewhere in this repo, I'm also down for that. Keep in mind, this is program written in rust language/

**Comment by cogito123** [comment]
I also reduced size of chunk_container.lua and river.lua, by dropping unnecessary index as suggested by @developer-8 and moved biter_texture to precomputed directory to make it more organized

**Comment by developer-8** [comment]
Original river variate water/land on 4 tiles. y for north side: -26, -25, -24, -23

After the PR is in the current state it variate on 4 tiles. y for north side: -28, -27, -26, -25
This also means that the river has expanded by 4 tiles.
image

**Comment by Ragnarok77-factorio** [comment]
Waiting for debate is over or PR updated before merging it.

**Comment by cogito123** [comment]
Thanks for double checking. I remember checking it myself, but perhaps I changed something to pattern itself after, which made it offset unexpectedly.

File: maps/biter_battles_v2/config.lua [hunk 1]
```diff
@@ -6,7 +6,7 @@ local bb_config = {
 	["south_side_team_name"] = "Team South",
 
 	--TERRAIN OPTIONS--
-	["border_river_width"] = 44,						--Approximate width of the horizontal impassable river separating the teams. (values up to 64)
+	["border_river_width"] = 46,						--Approximate width of the horizontal impassable river separating the teams. (values up to 64)
```
Comment (cogito123) [inline maps/biter_battles_v2/config.lua:9]: I changed only this

**Comment by developer-8** [comment]
`[ERROR] ...temp/currently-playing/maps/biter_battles_v2/terrain.lua:170: attempt to perform arithmetic on local 'seed' (a nil value) in .../temp/currently-playing/maps/biter_battles_v2/main.lua:115`
probably caused by no seed passed to function
 https://github.com/Factorio-Biter-Battles/Factorio-Biter-Battles/blob/76d52922d3f98f168914ccba1fc8c275b695c898/maps/biter_battles_v2/terrain.lua#L699

**Comment by cogito123** [comment]
Oh, you could totally landfill entire river with that... Fixed.
I'm not sure how, but is there some way to write automated tests?

**Comment by Ragnarok77-factorio** [comment]
Thanks for quickfix and info @cogito123 and @developer-8 . server updated for main, should be fine there, for captain it will be another time

**Comment by Ragnarok77-factorio** [comment]
No idea for tests.

File: maps/biter_battles_v2/config.lua (hunk 1 above)
Comment (developer-8) [reply maps/biter_battles_v2/config.lua:9]: image
After the PR is in the current state it variate on 4 tiles. y for north side: -27, -26, -25, -24
Original was y for north side: -26, -25, -24, -23

**Comment by Ragnarok77-factorio** [comment]
is it fine now @developer-8 or anything else to add?

File: maps/biter_battles_v2/config.lua (hunk 1 above)
Comment (developer-8) [inline maps/biter_battles_v2/config.lua:9]:
```suggestion
	["border_river_width"] = 44,						--Approximate width of the horizontal impassable river separating the teams. (values up to 64)
```
image
with this change river variates as in original at  y for north side: -26, -25, -24, -23

Comment (cogito123) [reply maps/biter_battles_v2/config.lua]: Done

Comment (Ragnarok77-factorio) [reply maps/biter_battles_v2/config.lua]: Resolving it as you fixed it from what you said.

**Review by developer-8 — APPROVED** [review]
Coast generation is acceptable.
Dont know about performance improvements.

File: maps/biter_battles_v2/terrain.lua [hunk 2]
```diff
@@ -257,17 +261,33 @@ end
...
-				surface.set_tiles({{name = "deepwater", position = pos}})
+	local seed = game.surfaces[global.bb_surface_name].map_gen_settings.seed
+	-- Stack allocated buffer with a capacity of 1024.
+	local tiles = chunk.buffer
+	local pos = { x = left_top_x }
+	local i = 1
+	for _ = 0, 31, 1 do
+		pos.y = left_top_y
+		for _ = 0, 31, 1 do
+			pos.y = pos.y + 1
```
Comment (developer-8) [inline maps/biter_battles_v2/terrain.lua:272]: Does it make sense to do the same elsewhere where iterate over chunk tiles,  or is it just for this case?

**Review by Ragnarok77-factorio — APPROVED** [review]
Sure, has been tested for quite some time now. code wise nothing to add (I didnt deeply check but seems fine..?)

**Pull request merged**